### ISSUE TITLE
How to handle ECONNRESET and EPIPE on streaming?

### ISSUE THREAD

**Issue by stepankuzmin** [koajs/koa#1089]
Consider this example

```js
const fs = require('fs');
const Koa = require('koa');

const app = new Koa();

app.use(async (ctx) => {
  ctx.body = fs.createReadStream('really-large-file');
});

app.listen(4000);
```

If user aborts request I'm getting either

```shell
  Error: read ECONNRESET
      at _errnoException (util.js:1024:11)
      at TCP.onread (net.js:618:25)
```

or

```shell
  Error: write EPIPE
      at _errnoException (util.js:1024:11)
      at WriteWrap.afterWrite [as oncomplete] (net.js:870:14)
```

What is the proper way to handle this type of errors?

**P.S.** I have no errors after request aborts with `express`

```js
const fs = require('fs');
const express = require('express');

const app = express();

app.get('/', (req, res) => {
  fs.createReadStream('really-large-file').pipe(res);
});

app.listen(4000);
```

**P.P.S.** I've tried

```js
app.use(async (ctx) => {
  fs.createReadStream('really-large-file').pipe(ctx.res);
  ctx.respond = false;
});
```

But it had no effect.

**Comment by dead-horse** [comment]
try [koa-onerror](https://github.com/koajs/onerror) to catch this error and handle it manually?

**Comment by jonathanong** [comment]
depends on how you want to handle the error. once the headers and the stream is already sent, there's not much you can do.

**Comment by GuillaumeCisco** [comment]
Ok just found out resolving this problem in my case is impossible as `event-stream` do not communicate data to the server. So the server will never know when an user refresh its tab. We have to use a websocket communication to handle this.

I discovered webpack-koa with https://github.com/webpack-contrib/webpack-hot-client which implements a websocket approch and should be better than using koa-error. I will try that ;)

**Comment by shellscape** [comment]
@GuillaumeCisco just thought I would mention: I wrote webpack-hot-client, and was researching the same ECONNRESET error in a separate project. your comment reminded me that I've already solved this problem in the past. Cheers.

**Comment by rvanmil** [comment]
I know this is an old issue but in case someone finds it and is unable to get things to work properly (koa-onerror did not work for me), here's how I got this to work (koa 2.7.0).
Looking at the `onerror` function here: https://github.com/koajs/koa/blob/8b4e2cd3bc6e165a0ea544686346cd79e437bc28/lib/context.js#L121 and here: https://github.com/koajs/koa/blob/8b4e2cd3bc6e165a0ea544686346cd79e437bc28/lib/context.js#L123-L128 once the headers are sent, nothing is done. But the error is emitted as an app-level error.
So I created a handler for the app-level `error` event:

```
// Catch and log Koa app-level errors
const app = new Koa()
app.on('error', (error) => {
  if (error && error.code === 'EPIPE') {
    logger.warn('Koa app-level EPIPE error.', { error })
  } else {
    logger.error('Koa app-level error', { error })
  }
})
```

And now I'm no longer seeing the `Error: write EPIPE` messages 😊
